sheerwater_benchmarking/utils/caching.py

```diff
@@ -444,38 +511,126 @@ def check_exists_postgres(table_name):
                            the tailnet and can see sheerwater-benchmarking-postgres.""")
 
 
-def write_to_postgres(pandas_df, table_name, overwrite=False):
+def write_to_postgres(df, table_name, overwrite=False, upsert=False, primary_keys=None, real_table_name=False):
     """Writes a pandas df as a table in the sheerwater postgres.
 
     Backends should eventually be flexibly specified, but for now
     we'll just hard code a connection to the running sheerwater database.
 
     Args:
-        pandas_df: A pandas dataframe
+        df: A pandas dataframe
         table_name: The table name to write to
         overwrite (bool): whether to overwrite an existing table
+        upsert (bool): whether to upsert into table
+        primary_keys (list(str)): column names of primary keys for upsert
+        real_table_name (bool): whether to hash the table name. Default false
     """
     # Get the postgres write secret
     pgwrite_pass = postgres_write_password()
 
-    new_table_name = postgres_table_name(table_name)
+    if real_table_name:
+        new_table_name = table_name
+    else:
+        new_table_name = postgres_table_name(table_name)
+
+    if upsert:
+        if primary_keys is None or not isinstance(primary_keys, list):
+            raise ValueError("Upsert may only be performed with primary keys specified as a list.")
+
+        if not isinstance(df, dd.DataFrame):
+            raise RuntimeError("Upsert is only supported by dask dataframes for parquet")
+
+        uri = f'postgresql://write:{pgwrite_pass}@34.59.163.82:5432/postgres'
+        engine = sqlalchemy.create_engine(uri)
+
+        with engine.begin() as conn:
+            # If the table doesn't exist then create it
+            if not conn.exec_driver_sql(f"""SELECT EXISTS (
+                                SELECT FROM information_schema.tables
+                                WHERE  table_schema = 'public'
+                                AND    table_name   = '{new_table_name}');
+                                """).first()[0]:
+
+                print("Postgres cache doesn't exist for upsert. Writing full table.")
+                if isinstance(df, pd.DataFrame):
```

Comment:
This is the same as the case below? (upsert false vs upsert true but no table) - can they share the same code path for this main logic?